### ISSUE TITLE
Actor scalability model

### ISSUE THREAD

**Issue by shafqatevo** [wasmCloud/wasmCloud#50]
Hi,

This is a project with huge potential!

The question I have is regarding the scalability of actors. Are the wasm actors designed to be instantiated by the millions in a typical host node, like Akka actors or Erlang processes? Like when serving requests, since the wasm actors are single-threaded, will one actor get instantiated per request or the actor will essentially serially serve requests or a pool of actors will serve the requests?

What’s the scalability model?

Also, what’s the current footprint size of a Hello World actor with essential capabilities including HTTP, in terms of size of the binary and runtime memory requirement (both taken by the wasm actor and the memory overhead of the host per such actor)?

Thanks!

**Comment by autodidaddict** [comment]
The footprint produced by a `.wasm` file varies a bit depending on which language produced it. Rust-compiled `.wasm` binaries are typically around **2MB** on disk including the cryptographic signature and embedded security claims. AssemblyScript can get down to about half that, depending on what you're doing. The HTTP server "hello world" sample that we use to baseline things is the "kvcounter" sample - an actor that accepts incoming HTTP requests, increments a counter value in a kvstore, and then returns a JSON payload over HTTP with the incremented value. 

The RAM overhead for this is 64KB (memory is allocated to wasm modules in 64KB pages), plus some negligible overhead for instantiated objects and pointers above that. I figure we probably consume ~90KB of RAM per idle actor plus the bytes used for its instructions (e.g. the 2MB file). When I've done capacity planning up to this point, I've been using **2.1MB** as a memory benchmark for each actor.

Obviously, at 2.1MB, these are far heavier than Erlang/Elixir actors, but that's because all of the instructions used by those actors are in RAM, and not part of the compiled host binary. It's the tradeoff we pay for having actors that can flow in and out of host runtimes, regardless of whether they're IoT, cloud, browser, or anywhere else.

The memory footprint can also vary a bit depending on whether you're using **JIT** compilation (the wasm instructions are converted into native instructions at load time) or interpreter. The interpreter model consumes pure RAM, while the JIT model actually increases the code page footprint of the host runtime process itself. wasmCloud supports an engine for JITting (wasmtime) and an engine for interpreting (wasm3), which you can choose to optimize your runtime performance for your use case.

Scalability _per actor_ is done with the `lattice` network, which you can think of as an OTP-like cluster, held together via message broker infrastructure rather than OTP's directly connected graph cluster. If you want to run 2 instances of the same actor, you'll currently need 2 runtime hosts within the same lattice (we don't _currently_ have the concept of an internal pool of the same actor, but also remember that our actors are stateless whereas OTP actors all "remember" their state via function parameters). If evidence shows this is a benefit, then we can drop an issue into the backlog to support "local actor pools".

In summary, you should think of the lattice as your blank canvas for scale - we tell the lattice (a collection of interconnected runtime hosts) to launch an actor via auction, and the winner starts the actor. you can auction the same actor `n` times to set its horizontal scaling property to `n`. The same is done for capability providers, which can also scale to `n` by running an instance of that provider on `n` hosts.

tl;dr - heavier than an Erlang actor, lighter than a microservice. More akin to an "always-hot" lambda.

**Comment by shafqatevo** [comment]
Hi @autodidaddict thanks so much for your elaborate and helpful answer!

I guess wasmCloud’s design goal is not high-concurrency then.

The 2MB image of the wasm that is kept in memory - is it because of interpreter model? Will AOT/JIT still require keeping the image resident in RAM?

I think I have seen in IOT circles or maybe Bytecode Alliance itself that people are saying 64KB page size is too large for embedded devices and it should be reduced or there should be a smaller option.

Other than the memory image and actor’s own page size, the per-actor overhead seems to be ~26KB, which is just 10x of Erlang’s ~2.4KB overhead (Akka’s per actor overhead is just ~350 bytes).

Of course, the sandboxing, portability, mobility of wasm is a different set of values.

**Comment by autodidaddict** [comment]
Yes when we get to embedded and constrained environments we need to look into running with smaller pages and smaller stacks.

When the 2mb image is JITted that is compiled into a library in memory which is then linked to the host so my gut says this takes up more RAM but I haven't proved this in tests yet.

**Comment by autodidaddict** [comment]
#54

**Comment by brooksmtownsend** [comment]
Discussion continued and tracked in #54
